bpf/bpf_lxc.c

```diff
@@ -1941,6 +1941,119 @@ int tail_ipv4_policy(struct __ctx_buff *ctx)
 	return ret;
 }
 
+static __always_inline bool
+ipv4_to_endpoint_is_hairpin_flow(struct __ctx_buff *ctx, struct iphdr *ip4, __u32 __maybe_unused src_sec_identity)
+{
+	struct ipv4_ct_tuple tuple = {};
+	struct lb4_backend *backend __maybe_unused;
+	struct ct_entry *entry;
+	struct ct_map *map;
+	int err, l4_off;
+	__be32 pod_ip, service_ip;
+	__be16 client_port, backend_port, service_port;
+
+	/* For reasons unknown, hairpinned packets don't have their IDs resolved and
+	 * are always seen as originating from world.
+	 */
+	if (src_sec_identity != WORLD_ID)
+		return false;
+
+	/* Extract the tuple from the packet so we can freely access addrs and ports.
+	 * All values are in network byte order.
+	 */
+	err = lb4_extract_tuple(ctx, ip4, ETH_HLEN, &l4_off, &tuple);
+	if (IS_ERR(err))
+		return false;
+
+	/* Two different types of packets need to be handled here. client->server and
+	 * server->client packets both pass through the to-endpoint path; presumably
+	 * they are NATed differently depending on the direction. We can observe both
+	 * service_ip:client_port -> pod_ip:service_port and
+	 * IPV4_LOOPBACK:backend_port -> pod_ip:client_port.
+	 */
+	if (tuple.saddr != IPV4_LOOPBACK) {
+		/* 
+		 * If the packet originates from a non-loopback address, it will look like:
```

Comment:
Nit: Comment style, here and elsewhere

```suggestion
		/* If the packet originates from a non-loopback address, it will look like:
```